Offload node events to cilium-operator and decrease number pod events received in cilium

aanm opened the pull request:
Once cilium is connected to the kvstore we can stop listenning for node
events from k8s and start listening for node events in the kvstore
instead.

A similar case goes to pod events where we can stop listening for all pod
events from k8s once we have connectivity with the kvstore and start listening
for pod events for pods running in the node where cilium is running.

~!!!! **BUG** !!!!: There's still a bug that I haven't figure out, the ip routes are not removed, 
from remaining nodes, once other nodes are deleted, it seems the cilium agents are ignoring
delete events from the kvstore.~

Signed-off-by: André Martins <andre@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7447)
<!-- Reviewable:end -->

aanm committed:
k8s: wait for certain subsystems to be synchronized with k8s

As we don't really need to wait for *all* k8s caches to be synchronized
we will wait for a subset of them.

Modified blockWaitGroupToSyncResources so we can call it multiple
times with the same resource name as well cancel the previous waiting
for caches to be synchronized. This will be useful in following PRs.

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -308,7 +306,7 @@ func (d *Daemon) initK8sSubsystem() chan struct{} {
 	go func() {
 		log.Info("Waiting until all pre-existing resources related to policy have been received")
-		d.k8sResourceSyncWaitGroup.Wait()
+		d.waitForCacheSync()
```
This seems to be a no-op and thus incorrect given this code:
```
func (d *Daemon) waitForCacheSync(resourceNames ...string) {
        for _, resourceName := range resourceNames {
                d.k8sResourceSyncedMu.RLock()
                c, ok := d.k8sResourceSynced[resourceName]
                d.k8sResourceSyncedMu.RUnlock()
                if !ok {
                        continue
                }
                <-c
        }
}
```

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -805,73 +806,84 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
+			go nodeController.Run(isConnected)
+			<-kvstore.Client().Connected()
+			close(isConnected)
+			d.k8sAPIGroups.removeAPI(k8sAPIGroupNodeV1Core)
```
I would put info messages here to indicate when we subscribe and unsubscribe to these events for clarity.

tgraf commented on daemon/daemon_main.go:
```diff
@@ -1184,7 +1184,7 @@ func runDaemon() {
 		}).Fatal("Unable to setup kvstore")
 	}
 
-	<-k8sCachesSynced
```
This does not make sense to me. Why are we not waiting anymore?

aanm committed:
pkg/kvstore: start watchers only when etcd is connected

There's no point in trying to watch for a resource is etcd is not
connected so we should start watching for a resource when cilium has
etcd connectivity.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/kvstore: add Connected and Disconnected functions

This allows other cilium subsystems to be aware when etcd is connected
or disconnected with the kvstore.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
kvstore/store: add method to sync keys directly in the KVStore

We might not want to keep a local state of all the keys store in the
kvstore so we need to add them directly in the kvstore. This will help
following up commits to make use of this functionality.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
operator: sync k8s nodes to the kvstore

To avoid all cilium-agents from listening for k8s nodes events,
which can cause extreme network I/O for a medium-large cluster
(500-1000 nodes), cilium operator will be the single entitiy that will
listen for such events and will only update the required fields for
cilium-agents to properly create the routes across nodes. Although all
of these cilium-agents will watch for node events directly from the
kvstore, it will be less data transfered from the kvstore to all of
these agents. We are to >= 400MiB/s if watching for k8s nodes events on
all cilium-agents vs <=25MiB/s if watching for node events of the
required fields from the kvstore.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s: watch for k8s node events until kvstore connectivity

As cilium requires the nodes routes to reach etcd, cilium will need
receive node events from k8s until cilium is connected with the kvstore.

Once that connectivity is successful cilium can stop watching for k8s
node events as all routes will be created based on kvstore watchers
which will be populated by cilium-operator.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s: watch for all pod events until kvstore connectivity

As cilium requires pod events from all cluster so it can populate the
ipcache correctly it needs to watch for pod events from k8s. However,
this is only required to be done until cilium has kvstore connectivity
for which it can cancel the pod watcher for all pods in the cluster and
start watching for pod events for the node where cilium is running.

Cilium still needs to watch for pod events in the node where it is
running as it needs to know if a pod changed its labels. For the pod
events belonging to other nodes ipcache entries will be synchronized
based on the kvstore watchers setup for this porpuse. If the kvstore
connectivity is lost, cilium will start watching for pod events for all
pods in the cluster again.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
examples/kubernetes: fix cilium-operator RBAC

cilium-operator has more permissions than the ones that actually need.
Re-wrote the RBAC for cilium-operator and add comments explaining the
reason it needs those permissions to access k8s API.

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -829,73 +831,102 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
+			d.k8sAPIGroups.removeAPI(k8sAPIGroupNodeV1Core)
+			// Create a new node controller when we are disconnected with the
+			// kvstore
+			<-kvstore.Client().Disconnected()
```
Anyone who uses Consul in conduit with Cilium will hit a `panic` now since `Disconnected` panics for Consul.
Maybe should maintain the old logic of having each agent subscribe to node events if Consul is being used, and specify that if users really want scale, they can use etcd.

tgraf replied:
We should just implement the `Disconnected()` function.

ianvernon commented on examples/kubernetes/1.10/cilium-containerd.yaml:
```diff
@@ -512,17 +512,31 @@ rules:
   resources:
-  - pods
-  - deployments
+  # to get k8s version and status
```
Should we backport these RBAC fixes?

tgraf commented on pkg/kvstore/etcd.go:
```diff
@@ -288,6 +288,47 @@ func (e *etcdClient) closeSession(leaseID client.LeaseID) {
+		select {
+		case <-ch:
+			// this case means we are disconnected as the e.session.Done() was
+			// closed. We will pooling until we have a new etcd session and
```
polling

tgraf commented on pkg/kvstore/etcd.go:
```diff
@@ -288,6 +288,47 @@ func (e *etcdClient) closeSession(leaseID client.LeaseID) {
+			// closed. We will pooling until we have a new etcd session and
+			// we are connected to the kvstore.
+			time.Sleep(100 * time.Millisecond)
+			goto getSession
```
It is very weird that you have different behavior in terms of when `Connected()` is blocking. When the initial session did not succeed yet it will return immediately and then close the channel. When the initial session was successful but closed, it will block until it is opened again. I would return the the channel and do the polling in a go routine as you do for consul.

tgraf commented on operator/k8s_node.go:
```diff
@@ -0,0 +1,99 @@
+						serNodes.Enqueue(func() error {
+							oldNode := k8s.ParseNode(oldNode, node.FromKubernetes)
+							newNode := k8s.ParseNode(newNode, node.FromKubernetes)
+							ciliumStore.DeleteLocalKey(oldNode)
```
This is not ideal. It will trigger a delete and add event via the kvstore. It will cause nodes to react to nodes disappearing and appearing and will result in routes get removed and re-added.

Add a function to update the key without relying on a local key. The node name is never changing so the key is *never* disappearing. It's always an update of an existing kvstore key.
